### PR TITLE
feat(react): react SDK with context provider

### PR DESCRIPTION
**Pull request by demonzoo**
## Summary

Provide SPA React SDK with Logto context provider and `useLogto` hook.

## Linear Issue Reference

[LOG-649](https://linear.app/silverhand/issue/LOG-649/jsreact)

## Testing

- [x] Passed all tests
- [x] Tested with react sample project. Works as expected. See screenshots below:

image

image

### CODE REVIEW

File: packages/react/src/components/ProtectedRoute/index.tsx [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import React, { useContext } from 'react';
+import { Route, RouteProps } from 'react-router-dom';
+
+import { LogtoContext } from '../../context';
+
+const ProtectedRoute = ({ ...routeProperties }: RouteProps) => {
+  const { isAuthenticated, isLoading, signIn } = useContext(LogtoContext);
+
+  if (isLoading) {
+    return <p>Loading...</p>;
```
Comment (gao-sun) [inline packages/react/src/components/ProtectedRoute/index.tsx:10]: this should be customizable?

Comment (demonzoo) [reply packages/react/src/components/ProtectedRoute/index.tsx:10]: Yes, we can allow passing a LoadingComponent through LogtoConfig, however, I'm not sure if we are going to do this in V1. Thoughts?

Comment (gao-sun) [reply packages/react/src/components/ProtectedRoute/index.tsx:10]: imo `LogtoConfig` should have no loading components. actually `ProtectedRoute` should be implemented by the developer not us.

File: packages/react/package.json [hunk 2]
```diff
@@ -18,19 +18,29 @@
     "report": "WITH_REPORT=true pnpm build"
   },
   "dependencies": {
-    "@logto/browser": "^0.1.0"
+    "@logto/browser": "^0.1.0",
+    "@silverhand/essentials": "^1.1.6",
+    "react": "^17.0.2",
```
Comment (wangsijie) [inline packages/react/package.json:23]: peerDependencies?

Comment (demonzoo) [reply packages/react/package.json:23]: ok

Comment (demonzoo) [reply packages/react/package.json]: We already have `react ^16.8.0` in the peerDependencies list

File: packages/react/src/components/ProtectedRoute/index.tsx (hunk 1 above)
Comment (demonzoo) [reply packages/react/src/components/ProtectedRoute/index.tsx]: Updated. PTAL

File: packages/react/src/provider.tsx [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  const [logtoClient, setLogtoClient] = useState(() => new LogtoClient(logtoConfig));
+  const [logtoContext, setLogtoContext] = useState<ExtendedLogtoContextProps>();
+  const [isLoading, setIsLoading] = useState(false);
+
+  const signIn = useCallback(
+    async (redirectUri: string) => {
+      setIsLoading(true);
+      await logtoClient.signIn(redirectUri);
+      setIsLoading(false);
+      setLogtoClient(logtoClient);
```
Comment (gao-sun) [inline packages/react/src/provider.tsx:21]: `logtoClient` should be a ref

File: packages/react/src/hooks/index.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import { useContext } from 'react';
+
+import { LogtoContext, LogtoContextProps } from '../context';
+
+export default function useLogto(): LogtoContextProps {
```
Comment (gao-sun) [inline packages/react/src/hooks/index.ts:5]: no need to wrap?

File: packages/react/package.json [hunk 5]
```diff
@@ -18,21 +18,27 @@
     "report": "WITH_REPORT=true pnpm build"
   },
   "dependencies": {
-    "@logto/browser": "^0.1.0"
+    "@logto/browser": "^0.1.0",
+    "@silverhand/essentials": "^1.1.6",
+    "react": "^17.0.2"
```
Comment (gao-sun) [inline packages/react/package.json:23]: react should be present in dev & peer deps

**Review by gao-sun — CHANGES_REQUESTED** [review]
title should be updated as well?

File: packages/react/package.json (hunk 5 above)
Comment (demonzoo) [reply packages/react/package.json:23]: done

File: packages/react/src/provider.tsx (hunk 3 above)
Comment (demonzoo) [reply packages/react/src/provider.tsx]: Updated.

File: packages/react/src/provider.tsx [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+import { LogtoContext, ExtendedLogtoContextProps } from './context';
+
+export type LogtoProviderProps = {
+  logtoConfig: LogtoConfig;
+  children?: React.ReactNode;
+};
+
+export const LogtoProvider = ({ logtoConfig, children }: LogtoProviderProps) => {
+  console.log('called!');
```
Comment (wangsijie) [inline packages/react/src/provider.tsx:12]: forget to remove?

**Review by wangsijie — APPROVED** [review]
overall LGTM

File: packages/react/src/provider.tsx (hunk 6 above)
Comment (demonzoo) [reply packages/react/src/provider.tsx:12]: Yeah... Sorry, my bad

File: packages/react/src/provider.tsx [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    return userInfo;
+  }, []);
+
+  const getAccessToken = useCallback(async () => {
+    setIsLoading(true);
+    const accessToken = await logtoClientReference.current.getAccessToken();
+    setIsLoading(false);
+
+    return accessToken;
+  }, []);
```
Comment (gao-sun) [inline packages/react/src/provider.tsx:48]: what if we call `fetchUserInfo` then `getAccessToken`? the `isLoading` will be set once one request has fulfilled, which is not expected. can be refactor in the next PR

File: packages/react/src/provider.tsx [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+  const getIdTokenClaims = useCallback(() => {
+    return logtoClientReference.current.getIdTokenClaims();
+  }, []);
+
+  useEffect(() => {
+    setLogtoContext({
+      logtoClient: logtoClientReference.current,
+      isAuthenticated: logtoClientReference.current.isAuthenticated,
```
Comment (gao-sun) [inline packages/react/src/provider.tsx:57]: looks like this won't be updated in time? there's no dependency in the hook. better expose a function?

Comment (demonzoo) [reply packages/react/src/provider.tsx]: Updated, good call

File: packages/react/package.json [hunk 9]
```diff
@@ -18,21 +18,27 @@
...
     "@jest/types": "^27.5.1",
+    "react": "^17.0.2",
     "@silverhand/eslint-config": "^0.9.1",
     "@silverhand/eslint-config-react": "^0.9.1",
     "@silverhand/ts-config": "^0.9.1",
     "@silverhand/ts-config-react": "^0.9.1",
+    "@testing-library/react-hooks": "^7.0.2",
     "@types/jest": "^27.4.1",
+    "@types/react": "^17.0.39",
+    "codecov": "^3.8.3",
```
Comment (gao-sun) [inline packages/react/package.json:34]: do we need this now?

Comment (simeng-li) [reply packages/react/package.json:34]: can b removed using github actions instead

**Review by simeng-li — APPROVED** [review]
LGTM

File: packages/react/package.json (hunk 9 above)
Comment (demonzoo) [reply packages/react/package.json]: Removed.

File: packages/react/src/provider.tsx (hunk 7 above)
Comment (demonzoo) [reply packages/react/src/provider.tsx]: Updated, please review

File: packages/react/src/hooks/index.ts [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+  const [isAuthenticated, setIsAuthenticated] = useState(
+    context.logtoClient?.isAuthenticated ?? false
+  );
+
+  const { logtoClient } = context;
+
+  const signIn = useCallback(
+    async (redirectUri: string) => {
+      if (!logtoClient) {
+        throw new Error('useLogto hook must be used inside LogtoProvider context');
```
[the remark this answers is not in the document]
Comment (demonzoo) [reply packages/react/src/hooks/index.ts]: Yes, updated

File: packages/react/src/hooks/index.ts [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+  fetchUserInfo: () => Promise<UserInfoResponse>;
+  getAccessToken: (resource?: string) => Promise<Nullable<string>>;
+  getIdTokenClaims: () => IdTokenClaims;
+  handleSignInCallback: (callbackUri: string) => Promise<void>;
+  signIn: (redirectUri: string) => Promise<void>;
+  signOut: (postLogoutRedirectUri: string) => Promise<void>;
+};
+
+export default function useLogto(): Logto {
+  const context = useContext(LogtoContext);
```
[the remark this answers is not in the document]
Comment (demonzoo) [reply packages/react/src/hooks/index.ts:21]: Updated

File: packages/react/src/context.tsx [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+export type LogtoContextProps = {
+  logtoClient?: LogtoClient;
+  loadingStackCount: number;
+  setLoadingStackCount: React.Dispatch<React.SetStateAction<number>>;
+};
+
+export const LogtoContext = createContext<LogtoContextProps>({
+  logtoClient: undefined,
+  loadingStackCount: 0,
+  setLoadingStackCount: (): never => {
```
Comment (gao-sun) [inline packages/react/src/context.tsx:13]: `loadingCount` would be fine

File: packages/react/src/context.tsx [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+    throw new Error('Must be used inside <LogtoProvider>');
```
Comment (gao-sun) [inline packages/react/src/context.tsx:14]: we can define the error builder in this file and reuse in `hooks`:

```ts
const throwContextError = () => throw new Error('...');
```

**Review by gao-sun — APPROVED** [review]

File: packages/react/src/context.tsx (hunk 13 above)
Comment (demonzoo) [reply packages/react/src/context.tsx:14]: [LOG-1854](https://linear.app/silverhand/issue/LOG-1854/declare-a-standardized-error-class-in-react-context)
Ticket created. Will do in a follow-up PR

File: packages/react/src/hooks/index.ts [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+  isLoading: boolean;
+  fetchUserInfo: () => Promise<UserInfoResponse>;
+  getAccessToken: (resource?: string) => Promise<Nullable<string>>;
+  getIdTokenClaims: () => IdTokenClaims;
+  handleSignInCallback: (callbackUri: string) => Promise<void>;
+  signIn: (redirectUri: string) => Promise<void>;
+  signOut: (postLogoutRedirectUri: string) => Promise<void>;
+};
+
+const notInProviderErrorMessage = 'useLogto hook must be used inside LogtoProvider context';
```
[the remark this answers is not in the document]
Comment (demonzoo) [reply packages/react/src/hooks/index.ts:18]: Ticket created: [LOG-1854](https://linear.app/silverhand/issue/LOG-1854/declare-a-standardized-error-class-in-react-context)

File: packages/react/src/context.tsx (hunk 12 above)
Comment (demonzoo) [reply packages/react/src/context.tsx]: Updated

**Pull request merged**